### PR TITLE
#5310 Fixes for attachments which contain emojis in their names

### LINKED ISSUE #5310: Unable to send attachment when filename contains emoji
When trying to send file with emoji in the filename I get error `Failed to send message due to: URIError: malformed URI sequence`:

```
malformed URI sequence
continuationEncodeChr@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/lib/emailjs/emailjs-mime-codec.js:661:28
continuationEncode@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/lib/emailjs/emailjs-mime-codec.js:712:31
MimeNode.prototype._buildHeaderValue/<@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/lib/emailjs/emailjs-mime-builder.js:552:27
MimeNode.prototype._buildHeaderValue@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/lib/emailjs/emailjs-mime-builder.js:549:46
MimeNode.prototype.build/<@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/lib/emailjs/emailjs-mime-builder.js:382:34
MimeNode.prototype.build@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/lib/emailjs/emailjs-mime-builder.js:371:23
MimeNode.prototype.build/<@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/lib/emailjs/emailjs-mime-builder.js:471:33
MimeNode.prototype.build@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/lib/emailjs/emailjs-mime-builder.js:469:30
Mime.encode@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/js/common/core/mime.js:245:25
SendableMsg/this.toMime@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/js/common/api/email-provider/sendable-msg.js:49:39
Gmail/this.msgSend@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/js/common/api/email-provider/gmail/gmail.js:122:51
ComposeSendBtnModule/this.attemptSendMsg@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/chrome/elements/compose-modules/compose-send-btn-module.js:219:58
ComposeSendBtnModule/this.doSendMsgs@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/chrome/elements/compose-modules/compose-send-btn-module.js:269:55
ComposeSendBtnModule/this.extractProcessSendMsg@moz-extension://98c7ae9f-e7d0-484e-b0ae-722ed95db7b9/chrome/elements/compose-modules/compose-send-btn-module.js:91:47

    at <async> encode (/extension/js/common/core/mime.ts:239:25)
    at <async> toMime (/extension/js/common/api/email-provider/sendable-msg.ts:190:18)
    at <async> msgSend (/extension/js/common/api/email-provider/gmail/gmail.ts:90:19)
    at <async> attemptSendMsg (/extension/chrome/elements/compose-modules/compose-send-btn-module.ts:221:27)
```

### LINKED ISSUE #5345: Attachment blocks are duplicated when filename contains non-alphanumeric symbols
Screenshot 2023-08-03 at 13 38 50

Gmail replaces emoji in filename with `` and it makes our regex for filename in [processAttachments](https://github.com/FlowCrypt/flowcrypt-browser/blob/8daed5fbe9008d246b0301b4f7952137eef1777e/extension/js/content_scripts/webmail/gmail-element-replacer.ts#L434) to fail.

Possible solution - remove emojis from filename in regex and ignore `` tags, then regex should work

### PR DESCRIPTION
**Pull request by sosnovsky**
This PR fixes issues for attachments which contain emojis in their filenames

close #5310
close #5345

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

File: extension/lib/emailjs/emailjs-mime-codec.js [hunk 1]
```diff
@@ -697,12 +697,22 @@
...
+
+                // fix for attachments with emoji in filenames
+                if (typeof Intl !== 'undefined' && typeof Intl.Segmenter === 'function') {
+                    // Intl.Segmenter() currently not available on Firefox
+                    // https://caniuse.com/mdn-javascript_builtins_intl_segmenter
+                    encodedStr = [...new Intl.Segmenter().segment(encodedStr)].map(x => x.segment)
+                } else {
+                    // regex from https://stackoverflow.com/a/69661174/3091318
+                    encodedStr = encodedStr.replace(/(?![*#0-9]+)[\p{Emoji}\p{Emoji_Modifier}\p{Emoji_Component}\p{Emoji_Modifier_Base}\p{Emoji_Presentation}]/gu, '')
+                }
```
Comment (martgil) [inline extension/lib/emailjs/emailjs-mime-codec.js:714]: This is nice - I tried to find a reference for regex in Kotlin but I do not see any relevant documentation/reference material that Kotlin supports \p{Emoji_Presentation} or similar, unlike javascript which has rich support with it. So with that, we might end up using emoji-unicode ranges which again may contain an incomplete range of emojis. 

@sosnovsky Do you probably allow if I fix both #5308 and #5309 by re-using the regex on line 713?

Comment (sosnovsky) [reply extension/lib/emailjs/emailjs-mime-codec.js:714]: Yeah, I tried to use the same regex on FES for allowing emojis, but unfortunately Kotlin doesn't support these `\p` keywords.

Let me finish this PR, then I'll check possible solution for #5308 and #5309 on FES side, and if there won't be a simple way to allow emojis and other symbols in recipient names - I'll notify you that you can work on stripping emojis from recipient names on browser extension side.

File: test/source/tests/compose.ts [hunk 2]
```diff
@@ -2283,6 +2283,22 @@ export const defineComposeTests = (testVariant: TestVariant, testWithBrowser: Te
...
+        const acctEmail = 'flowcrypt.compatibility@gmail.com';
+        await BrowserRecipe.setupCommonAcctWithAttester(t, browser, 'compatibility', {
+          attester: { includeHumanKey: true },
+        });
+        const subject = `Test Sending Message With Attachment Which Contains Emoji in Filename ${Util.lousyRandom()}`;
+        const composePage = await ComposePageRecipe.openStandalone(t, browser, acctEmail);
+        await ComposePageRecipe.fillMsg(composePage, { to: 'human@flowcrypt.com' }, subject);
+        const fileInput = (await composePage.target.$('input[type=file]')) as ElementHandle<HTMLInputElement>;
+        await fileInput!.uploadFile('test/samples/attac👍hment!🔸.txt');
+        await ComposePageRecipe.sendAndClose(composePage);
```
Comment (ioanmo226) [inline test/source/tests/compose.ts:2298]: I manually tested using the `attac👍hment!🔸.txt.`.
Sending the email went smoothly, but on the receiver's end, both the filename and file content were incorrect. (file content seems malformed)

Plus, I think it's better to test if sent attachment file name & content is correct in ui test.

image

Comment (sosnovsky) [reply test/source/tests/compose.ts:2298]: Hmm, same for me, I think earlier it was sent with the correct attachment name, will check

Comment (sosnovsky) [reply test/source/tests/compose.ts:2298]: It works in Firefox, but sends incorrect filename in Chrome, will fix it

Comment (sosnovsky) [reply test/source/tests/compose.ts:2299]: @ioanmo226 I fixed this issue and also updated test to check attachment filename of the sent message

**Review by ioanmo226 — APPROVED** [review]
Looks good 👍

**Pull request merged**